### PR TITLE
Remove in-tree support for ControlPlane exposure components and rely on extension controllers

### PR DESCRIPTION
**Pull request by svetlinas**
**What this PR does / why we need it**:
Create/Delete new ControlPlane resource with `purpose` set to `exposure`. It triggers the new control plane controllers for the exposure settings. 

**Which issue(s) this PR fixes**:
gardener/gardener-extensions#233 

**Special notes for your reviewer**:
:warning: For local development you need to update your provider extension charts and use at least version `0.10.0-dev-68668b0d1518f8e695ec09049931a4409e965906`. We'll create a new release of the gardener-extensions repository before releasing gardener.

**Release note**:

```action operator
:warning: Gardener does no longer have in-tree supported for managing provider-specific control plane configuration and deployments of additional provider-specific components relevant for exposing the control plane (like aws-lb-readvertiser). Instead, it now relies on extension controllers to deploy the respective components for the specific provider they have been developed for, e.g. AWS, Azure, GCP, etc. (see [GEP-1 on extensibility](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md)). We have prepared the implementation of the [Alicloud provider](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-alicloud), [AWS provider](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-aws), [Azure provider](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-azure), [GCP provider](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-gcp), [OpenStack provider](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-openstack), and [Packet provider](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-packet). After updating Gardener you need to update your existing `ControllerRegistration` resources to make these extension controllers known to Gardener (otherwise none of your shoots will be able to get reconciled anymore). Please find example `ControllerRegistration` resources [here (Alicloud)](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-alicloud/example/controller-registration.yaml), [here (AWS)](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-aws/example/controller-registration.yaml), [here (Azure)](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-azure/example/controller-registration.yaml), [here (GCP)](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-gcp/example/controller-registration.yaml), [here (OpenStack)](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-openstack/example/controller-registration.yaml), and [here (Packet)](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/provider-packet/example/controller-registration.yaml). To get information about more details please walk through [these](https://github.com/gardener/gardener/tree/master/docs/extensions) and [these documents](https://github.com/gardener/gardener/tree/master/docs/extensions/controlplane-exposure.md).
:warning: Please make sure you upgrade your extension controllers to at least version [0.10.0] before upgrading Gardener.
```

### CODE REVIEW

File: pkg/controllermanager/controller/shoot/shoot_control_delete.go [hunk 1]
```diff
@@ -291,6 +291,16 @@ func (c *Controller) runDeleteShootFlow(o *operation.Operation) *gardencorev1alp
 			Fn:           flow.TaskFn(botanist.WaitUntilManagedResourcesDeleted).Timeout(10 * time.Minute),
 			Dependencies: flow.NewTaskIDs(deleteManagedResources),
 		})
+		destroyCloudSpecificControlPlane = g.Add(flow.Task{
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_control_delete.go]: I think we should delete it after the kube-apiserver deployment is deleted. It's relevant for the kube-apiserver exposure, and if it is gone then we can safely delete all exposure-relevant components.

File: pkg/controllermanager/controller/shoot/shoot_control_reconcile.go [hunk 2]
```diff
@@ -181,15 +181,21 @@ func (c *Controller) runReconcileShootFlow(o *operation.Operation, operationType
 			Fn:           flow.TaskFn(botanist.WaitUntilKubeAPIServerReady).SkipIf(o.Shoot.HibernationEnabled),
 			Dependencies: flow.NewTaskIDs(deployKubeAPIServer),
 		})
+		// TODO all the methods are now duplicated for the 2 control plane resources
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_control_reconcile.go]: What do you mean by "duplicated"?

File: pkg/operation/botanist/controlplane.go [hunk 3]
```diff
@@ -708,17 +708,75 @@ func (b *Botanist) DeployControlPlane(ctx context.Context) error {
...
+const controlPlaneExposureSuffix = "-exposure"
+
+// DeployCloudSpecificControlPlane creates the `ControlPlane` extension resource with purpose `exposure` in the shoot
+// namespace in the seed cluster. Gardener waits until an external controller did reconcile the cluster successfully.
+func (b *Botanist) DeployCloudSpecificControlPlane(ctx context.Context) error {
+	var (
+		cp = &extensionsv1alpha1.ControlPlane{
+			ObjectMeta: metav1.ObjectMeta{
+				// TODO the name of the resource can be different - the name of the seed?
+				// TODO by adding suffix we can understand quickly with control plane for what scenario is
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]: I like the suffix approach. The current name is fine, I guess :)

File: pkg/operation/botanist/controlplane.go [hunk 4]
```diff
@@ -708,17 +708,75 @@ func (b *Botanist) DeployControlPlane(ctx context.Context) error {
 	})
 }
 
+const controlPlaneExposureSuffix = "-exposure"
+
+// DeployCloudSpecificControlPlane creates the `ControlPlane` extension resource with purpose `exposure` in the shoot
+// namespace in the seed cluster. Gardener waits until an external controller did reconcile the cluster successfully.
+func (b *Botanist) DeployCloudSpecificControlPlane(ctx context.Context) error {
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]:
```suggestion
func (b *Botanist) DeployControlPlaneExposure(ctx context.Context) error {
```

(adapt other names accordingly, please)

**Review by rfranzke — CHANGES_REQUESTED** [review]
Please also remove:

* https://github.com/gardener/gardener/blob/master/charts/images.yaml#L33-L36
* https://github.com/gardener/gardener/tree/master/charts/seed-controlplane/charts/aws-lb-readvertiser
* https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/secrets.go#L185-L204
* https://github.com/gardener/gardener/blob/master/pkg/operation/common/types.go#L212-L214
* https://github.com/gardener/gardener/blob/master/pkg/operation/common/types.go#L460-L462
* Everything related to the botanist health check that checks the `aws-lb-readvertiser` deployment (`pkg/operation/botanist/health_check.go`)
* Rename https://github.com/gardener/gardener/blob/master/pkg/utils/kubernetes/kubernetes_test.go#L279-L280

**Comment by svetlinas** [comment]
> * Rename https://github.com/gardener/gardener/blob/master/pkg/utils/kubernetes/kubernetes_test.go#L279-L280

Why should this tests be renamed? They are validating a method ValidDeploymentContainerImageVersion(...) which I do not think is related to the AWS components (the method is not used in gardener/gardener project). The test data is named "aws-lb-readvertiser". What about rename the test data?

**Comment by rfranzke** [comment]
Yes, I meant renaming the test data, sorry for being not accurate enough. :)

**Review by rfranzke — COMMENTED** [review]
Could you please add a document (`docs/extensions/controlplane-exposure.md`) (similar to https://github.com/gardener/gardener/blob/master/docs/extensions/controlplane.md) that explains the purpose of this resource, why it exists, what it can be used for, how it works, and a maybe a reference to the AWS implementation as an example?

File: docs/extensions/controlplane-exposure.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+Some Kubernetes clusters require an additional deployments required by the seed cloud provider in order to work properly, e.g. AWS Load Balancer Readvertiser.
+Before using ControlPlane resources with purpose `exposure` Gardener was having different Helm charts for the deployments for the various providers.
+Now, Gardener commissions an external, provider-specific controller to take over this task.
+
+## Which control plane resources are required?
+
+As mentioned in the [controlplane](controlplane.md) document Gardener shall not deploy any other provider-specific component.
+Instead, it creates a `ControlPlane` CRD with purpose `exposure` that should be picked up by provider extensions.
+Its purpose is to trigger the deployment of such provider-specific components in the shoot namespace in the seed cluster that are need for the seed cluster.
```
Comment (rfranzke) [inline docs/extensions/controlplane-exposure.md]:
```suggestion
Its purpose is to trigger the deployment of such provider-specific components in the shoot namespace in the seed cluster that are needed to expose the kube-apiserver.
The shoot cluster's kube-apiserver are exposed via a `Service of type `LoadBalancer`. This load balancer is part of the seed provider's infrastructure. As the seed provider might differ from the shoot provider (you may run the control plane of an Azure shoot in a GCP seed) it's the seed provider extension controller that should act on the `ControlPlane` resources with purpose `exposure`.
```

File: docs/extensions/controlplane-exposure.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+spec:
+  type: aws
+  purpose: exposure
+  region: europe-west1
+  secretRef:
+    name: cloudprovider
+    namespace: shoot--foo--bar
+```
+
+The `.spec.secretRef` contains a reference to the provider secret pointing to the account that shall be used for the shoot cluster. The `.spec.region` contains the region of the seed cluster.
```
Comment (rfranzke) [inline docs/extensions/controlplane-exposure.md]:
```suggestion
The `.spec.secretRef` contains a reference to the provider secret pointing to the account that shall be used for the shoot cluster.
It is most likely not needed, however, still added for some potential corner cases.
If you don't need it then just ignore it.
The `.spec.region` contains the region of the seed cluster.
```

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**